components/sup/src/ctl_gateway/server.rs

```diff
@@ -453,36 +460,97 @@ struct SrvState {
 }
 
-/// Start a new thread which will run the CtlGateway server.
-///
-/// New connections will be authenticated using `secret_key`. Messages from the main thread
-/// will be sent over the channel `mgr_sender`.
-pub async fn run(listen_addr: SocketAddr, secret_key: String, mgr_sender: MgrSender) {
-    let state = SrvState { secret_key,
-                           mgr_sender };
-    let state = Arc::new(Mutex::new(state));
-    let mut listner =
-        TcpListener::bind(&listen_addr).await
-                                       .expect("Could not bind ctl gateway listen address!");
-    let mut incoming = listner.incoming();
-    while let Some(tcp_stream) = incoming.next().await {
-        match tcp_stream {
-            Ok(tcp_stream) => {
-                let addr = match tcp_stream.peer_addr() {
-                    Ok(addr) => addr,
-                    Err(e) => {
-                        debug!("Client peer address not available from socket, err {}", e);
-                        continue;
+pub(crate) struct CtlGatewayServer {
+    pub(crate) listen_addr:         SocketAddr,
+    pub(crate) secret_key:          String,
+    pub(crate) mgr_sender:          MgrSender,
+    pub(crate) server_certificates: Option<Vec<Certificate>>,
+    pub(crate) server_key:          Option<PrivateKey>,
+    pub(crate) client_certificates: Option<RootCertStore>,
+}
+
+impl CtlGatewayServer {
+    /// Start a new thread which will run the CtlGateway server.
+    ///
+    /// New connections will be authenticated using `secret_key`. Messages from the main thread
+    /// will be sent over the channel `mgr_sender`.
+    pub async fn run(self) {
+        let Self { listen_addr,
+                   secret_key,
+                   mgr_sender,
+                   server_certificates,
+                   server_key,
+                   client_certificates, } = self;
+
+        let state = SrvState { secret_key,
+                               mgr_sender };
+        let state = Arc::new(Mutex::new(state));
+        let mut listener =
+            TcpListener::bind(&listen_addr).await
+                                           .expect("Could not bind ctl gateway listen address!");
+
+        // TLS configuration
+        let maybe_tls_config = if let Some(server_key) = server_key {
+            let mut tls_config = if let Some(client_certificates) = client_certificates {
+                debug!("Upgrading ctl-gateway to TLS with client authentication");
+                TlsServerConfig::new(AllowAnyAuthenticatedClient::new(client_certificates))
+            } else {
+                debug!("Upgrading ctl-gateway to TLS");
+                TlsServerConfig::new(NoClientAuth::new())
+            };
+            tls_config.set_single_cert(server_certificates.unwrap_or_default(), server_key)
+                      .expect("Could not set certificate for ctl gateway!");
+            Some(Arc::new(tls_config))
+        } else {
+            None
+        };
```

Comment:
Similarly, it might be better for the CtlGatewayServer to be able to take an optional `TlsServerConfig`, rather than having the logic for creating that embedded here.

Reply:
Addressed in https://github.com/habitat-sh/habitat/pull/7923/commits/8643bcfe2389395cf6078ba2240513202e915edd